torch/distributed/_spmd/gm_transformation.py

```diff
@@ -0,0 +1,131 @@
+            orig_placeholders.append(node)
+        else:
+            move_nodes.append(node)
+    orig_placeholders_set = set(orig_placeholders)
+
+    subgraph: torch.fx.Graph = torch.fx.Graph()
+    node_mapping: Dict[torch.fx.Node, torch.fx.Node] = {}
+    attrs = {}
+    for i, p in enumerate(orig_placeholders):
+        node_mapping[p] = subgraph.placeholder(name=f"placeholder_{i}")
+
+    for node in move_nodes:
+        if node.op == OP.GET_ATTR:
+            attrs[node.target] = getattr(gm, node.target)
+        args, kwargs = tree_map_only(
+            fx.Node, lambda n: node_mapping[n], (node.args, node.kwargs)
+        )
+        node_mapping[node] = subgraph.create_node(
+            op=node.op, target=node.target, args=args, kwargs=kwargs
+        )
+        node_mapping[node].meta = node.meta
+
+    placeholders: List[torch.fx.Node] = []
+    for placeholder in orig_placeholders:
+        new_placeholder = node_mapping[placeholder]
+        if len(new_placeholder.users) == 0:
+            # Remove unused placeholders from the subgraph. This is required as
+            # the `train_step()` has module and optimizer as the inputs which
+            # cannot be lowered to Inductor.
+            subgraph.erase_node(new_placeholder)
+        else:
+            placeholders.append(placeholder)
+    output_args = tuple(
+        node_mapping[n] for n in orig_output_args if n not in orig_placeholders_set
+    )
+    subgraph.output(result=output_args)
+    sub_gm = torch.fx.GraphModule(root=attrs, graph=subgraph)
+    gm.subgraph = InductorWrapper(sub_gm, enable_cudagraphs)
+    with gm.graph.inserting_after(move_nodes[-1]):
+        subgraph_call = gm.graph.create_node(
+            op=OP.CALL_MODULE, target="subgraph", args=tuple(placeholders)
+        )
+
+    # Redistribute the output from the subgraph to the original output.
+    output_idx = 0
+    for i, node in enumerate(orig_output_args):
+        if node in orig_placeholders_set:
+            continue
+        with gm.graph.inserting_after(subgraph_call):
+            new_node = gm.graph.call_function(
+                operator.getitem, (subgraph_call, output_idx)
+            )
+            output_idx += 1
+        orig_output_args[i] = new_node
+    assert output_idx == len(output_args)
+    gm.graph.erase_node(output)
+    gm.graph.output(result=orig_output_args)
+
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    return gm
+
+
+class GraphModuleTransformation:
+    def __init__(
+        self,
+        num_iters: int,
+        enable_inductor: bool = False,
+        enable_cudagraphs: bool = False,
+    ) -> None:
+        self.num_iters = num_iters
+        self.enable_inductor = enable_inductor
+        self.enable_cudagraphs = enable_cudagraphs
+
+    def __call__(self, gm: fx.GraphModule) -> Callable:
+        iter_gm = IterGraphModule(gm)
+        iter_gm.freeze_cross_iter_movement()
+        iter_gm.setup(self.num_iters)
```

Comment:
could these two methods be called automatically in `IterGraphModule` constructor?

Reply:
`freeze_cross_iter_movement()` can be called after IterGraphModule is integrated with Inductor.  `setup(num_iters)` is not possible to be removed at this moment.

Some examples that extra function call or extra argument to `forward()` is required other than simply arguments to the constructor:

1. Multiple epochs. For multiple epoch case, we need some signal to indicate that the current training is over and a reset will be needed for the next training loop.
2. Data depleted, similar to example 1. We need a way for users to let IterGraphModule know that training is over.

Will need some discussion about what's the easiest way for users to inform this information. Right now, we have to keep this API.